### PR TITLE
Make analyzer mvid part of our checksum

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
In order to support reloading of generators (which will only be supported in OOP due to this needing .net core capabilities), we need to know when the generator's containing analyzer-reference has actually changed so that we can ref the corresponding data on the OOP side, create a *new* AnalyzerReference there, and have that new reference be loaded properly in its corresponding shadow copied location.

It turns out the first part of this is actually *exceedingly* easy to do.  First, we just mix the MVID of the analzyer reference into our checksums.  That way if the mvid actually changes on disk, you get a new checksum.  Next, we take advantage of what we learned in https://github.com/dotnet/roslyn/pull/74720, namely that we are always using a 'shadow copying loader' everywhere in the IDE code.  This shadow copying loader is already brilliant for this sort of scenario.  That's because it already uses the MVID *already* of an Assembly Reference to copy the files to their own custom location (see https://github.com/dotnet/roslyn/blob/6696095f5aa51290707b5d8518b861ee0bab7884/src/Compilers/Core/Portable/DiagnosticAnalyzer/ShadowCopyAnalyzerAssemblyLoader.cs#L132-L147).

### CODE REVIEW

File: src/Workspaces/Core/Portable/Serialization/SerializerService_Reference.cs [hunk 1]
```diff
@@ -70,6 +70,7 @@ public static Checksum CreateChecksum(AnalyzerReference reference, CancellationT
                 case AnalyzerFileReference file:
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Serialization/SerializerService_Reference.cs:70]: view with whitespace off.

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski @ToddGrun ptal.

**Comment by ToddGrun** [comment]
>         var project2 = project1.RemoveAnalyzerReference(analyzerReference1).AddAnalyzerReference(analyzerReference2);

nit: might be nice to verify after the remove (but before the add) too

---
Refers to: src/Workspaces/CoreTest/SolutionTests/SolutionWithSourceGeneratorTests.cs:901 in 32957c8. [](commit_id = 32957c85508eefe283794b83ae9ee48991783f63, deletion_comment = False)

**Review by ToddGrun — APPROVED** [review]
:shipit:

File: src/Workspaces/Core/Portable/Serialization/SerializerService_Reference.cs [hunk 2]
```diff
@@ -504,13 +525,26 @@ private static void WriteUnresolvedAnalyzerReferenceTo(AnalyzerReference referen
...
+    {
+        try
+        {
+            return AssemblyUtilities.ReadMvid(file.FullPath);
+        }
+        catch
+        {
+            // We have a reference but the file the reference is pointing to might not actually exist on disk. In that
+            // case, rather than crashing, we will handle it gracefully.
+            return Guid.Empty;
```
Comment (jaredpar) [inline src/Workspaces/Core/Portable/Serialization/SerializerService_Reference.cs:544]: Another case is that a user has provided an analyzer file reference that isn't a valid managed assembly. That does happen from time to time. Would also trigger this code path.

**Pull request merged**